### ISSUE TITLE
New export rules broke backend specific FFI

### ISSUE THREAD

**Issue by bamboo** [idris-lang/Idris-dev#2960]
a14e4edea00295b6b73a2515264e88457396453a broke the `cil` backend FFI (and seems to have broken `python`'s  as well).

Calling `Idris.ElabDecls.elabMain` on a Idris function declared as `main : CIL_IO ()` now fails with `(toplevel:Elaborating run__IO arg __pi_arg: CantUnify True (CIL.FFI.CIL_IO Unit,Just (SourceTerm Main.main)) (IO' {ffi503} Unit,Just ExpectedType) CantUnify True (CIL.FFI.CIL_IO Unit,Nothing) (IO' {ffi503} Unit,Nothing)  in [({ffi503},FFI)] 0 in [({ffi503},FFI)] 0)`. 

This is what [`CIL_IO`](https://github.com/bamboo/idris-cil/blob/master/rts/CIL/FFI.idr#L96) looks like:

```Idris
%access public export

FFI_CIL : FFI
FFI_CIL = MkFFI CIL_Types CILForeign String

CIL_IO : Type -> Type
CIL_IO a = IO' FFI_CIL a
```

I've also tried to explicitly declare them as `public export` but the outcome was the same.

Is there a way to fix it on the backend side?

**Comment by edwinb** [comment]
I had to change the FFI descriptors in IO.idr to get the JS and C FFIs working with the new export rules. It's likely you need to do the same thing - this is unlikely (though not impossible) to be a problem with the export rules themselves. What's CIL_Types here? Is it exported as public, for example?

**Comment by bamboo** [comment]
CIL_Types is defined in the same [module](https://github.com/bamboo/idris-cil/blob/master/rts/CIL/FFI.idr#L5) marked with `%access public export`. I did try to explicitly declare every single definition in that module with `public export` but the error was the same.

**Comment by edwinb** [comment]
Is `main` exported? It isn't treated specially, so if you're importing a module with a `main` the elaborator won't know about it.

The REPL and Compiler get around this by loading ibcs with the IBC_REPL flag set which means that loading the IBC exports everything.

**Comment by bamboo** [comment]
`main` is not exported but the error is the same if I add `%access public export` to those modules as well.

The way idris-codegen-cil loads modules is very similar to how idris-codegen-javascript does it: 

```
cilMain :: Opts -> Idris ()
cilMain opts = do
  elabPrims
  _ <- loadInputs (inputs opts) Nothing
  mainProg <- elabMain
  ir <- compile (Via "cil") (output opts) (Just mainProg)
  runIO $ codegenCil ir
```

`elabMain` fails on `main : CIL_IO ()` and not on `main : IO ()`. I wonder if it has anything to do with the fact that `IO` (and `JS_IO` for that matter) are declared in a `%unqualified` module?

**Comment by bamboo** [comment]
Fixed by fdfd99e3cad2685720282598e6e8b7da911a012e.
